### ISSUE TITLE
PDL 2.038 build fails under Strawberry Perl 5.28.0, PDL edition

### ISSUE THREAD

**Issue by shawnlaffan** [PDLPorters/pdl#304]
Build log is below.  

FWIW, 5.30.0 seems to work.  Installation is incomplete at the time of writing, but well past the fail point for 5.28.  I will update if it does not pass. 

```
perl -v

This is perl 5, version 28, subversion 0 (v5.28.0) built for MSWin32-x64-multi-thread
```

```
cpanm (App::cpanminus) 1.7044 on perl 5.028000 built for MSWin32-x64-multi-thread
Work directory is C:\BERRYB~1\528~1.0_6\data/.cpanm/work/1618819138.29432
You have make C:\berrybrew\5.28.0_64_PDL\c\bin\gmake.exe
You have LWP 6.34
Falling back to Archive::Tar 2.38
Searching PDL () on cpanmetadb ...
--> Working on PDL
Fetching http://www.cpan.org/authors/id/E/ET/ETJ/PDL-2.038.tar.gz
-> OK
Unpacking PDL-2.038.tar.gz
Entering PDL-2.038
Checking configure dependencies from META.json
Checking if you have ExtUtils::ParseXS 3.01 ... Yes (3.39)
Checking if you have Carp 1.2 ... Yes (1.50)
Checking if you have ExtUtils::Depends 0.402 ... Yes (0.405)
Checking if you have File::Path 0 ... Yes (2.17)
Checking if you have ExtUtils::MakeMaker 7.12 ... Yes (7.46)
Checking if you have Devel::CheckLib 1.01 ... Yes (1.14)
Configuring PDL-2.038
Running Makefile.PL
Checking if your kit is complete...
Looks good
        1 file(s) copied.
Warning (mostly harmless): No library found for 
Use of uninitialized value in concatenation (.) or string at ./Makefile.PL line 20.

Not building PDL::Graphics::PGPLOT, PGPLOT not installed

No support for PDL::Graphics::IIS on win32 platform yet. Will skip build process

ExtUtils::F77: Version 1.22

Curses capable library not found, not building PDL::IO::Browser

Error loading Alien::HDF4: ' *** Can't load dependency information for Alien::HDF4:

Can't locate Alien/HDF4/Install/Files.pm in @INC (you may need to install the Alien::HDF4::Install::Files module) (@INC contains: C:/berrybrew/5.28.0_64_PDL/perl/site/lib/MSWin32-x64-multi-thread C:/berrybrew/5.28.0_64_PDL/perl/site/lib C:/berrybrew/5.28.0_64_PDL/perl/vendor/lib C:/berrybrew/5.28.0_64_PDL/perl/lib C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038 C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/IO .) at C:/berrybrew/5.28.0_64_PDL/perl/vendor/lib/ExtUtils/Depends.pm line 185.


'

Skipping build of PDL::IO::HDF.
Loaded ExtUtils::F77 version 1.22
Found compiler gfortran.exe
ExtUtils::F77: Using system=MSWin32 compiler=<undefined>
         ExtUtils::F77: Unable to guess and/or validate system/compiler configuration
         ExtUtils::F77: Will try system=Generic Compiler=GFortran
ExtUtils::F77: Validating "-LC:/berrybrew/5.28.0_64_PDL/c/bin/../lib/gcc/x86_64-w64-mingw32/7.1.0" -L/usr/lib -lgfortran -lquadmath -lm   [ok]
ExtUtils::F77: Compiler: gfortran
ExtUtils::F77: Cflags: -O
Compiling the test Fortran program...
Executing the test program...
Congratulations you seem to have a working f77!
ExtUtils::F77: Using system=MSWin32 compiler=<undefined>
         ExtUtils::F77: Unable to guess and/or validate system/compiler configuration
         ExtUtils::F77: Will try system=Generic Compiler=GFortran
ExtUtils::F77: Validating "-LC:/berrybrew/5.28.0_64_PDL/c/bin/../lib/gcc/x86_64-w64-mingw32/7.1.0" -L/usr/lib -lgfortran -lquadmath -lm   [ok]
ExtUtils::F77: Compiler: gfortran
ExtUtils::F77: Cflags: -O
Compiling the test Fortran program...
Executing the test program...
Congratulations you seem to have a working f77!
Generating a gmake-style Makefile
Writing Makefile for PDL
Writing MYMETA.yml and MYMETA.json
Writing Basic/Core/Config.pm
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have Storable 1.03 ... Yes (3.15)
Checking if you have Test::Exception 0 ... Yes (0.43)
Checking if you have List::MoreUtils 0 ... Yes (0.428)
Checking if you have File::Path 0 ... Yes (2.17)
Checking if you have Pod::Select 0 ... Yes (1.63)
Checking if you have Text::Balanced 1.89 ... Yes (2.03)
Checking if you have Scalar::Util 0 ... Yes (1.50)
Checking if you have File::Map 0.57 ... Yes (0.67)
Checking if you have Inline 0.83 ... Yes (0.86)
Checking if you have Convert::UU 0 ... Yes (0.5201)
Checking if you have Data::Dumper 2.121 ... Yes (2.173)
Checking if you have Pod::Parser 0 ... Yes (1.63)
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.46)
Checking if you have OpenGL 0.70 ... Yes (0.70)
Checking if you have Term::ReadKey 2.34 ... Yes (2.38)
Checking if you have IPC::Cmd 0.72 ... Yes (1.02)
Checking if you have Filter::Simple 0.88 ... Yes (0.94)
Checking if you have Test::Warn 0 ... Yes (0.35)
Checking if you have CPAN::Meta 2.120900 ... Yes (2.150010)
Checking if you have Inline::C 0.62 ... Yes (0.81)
Checking if you have Module::Compile 0.23 ... Yes (0.37)
Checking if you have Test::Deep 0 ... Yes (1.130)
Checking if you have Filter::Util::Call 0 ... Yes (1.58)
Checking if you have autodie 0 ... Yes (2.29)
Building and testing PDL-2.038
cp Bugs.pod blib\lib/PDL/Bugs.pod
gmake[1]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic'
gmake[2]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Pod'
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MPod::Select -e "podselect('../Gen/Inline/Pdlpp.pm');" > PP-Inline.pod
gmake[2]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Pod'
gmake[2]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Gen'
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" PP/dump.pp > PP/Dump.pm.tmp
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e mv -- PP/Dump.pm.tmp PP/Dump.pm
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MPod::Select -e "podselect('pptemplate');" > pptemplate.pod
gmake[3]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Gen/Inline'
gmake[3]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Gen/Inline'
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e cp -- pptemplate ..\..\blib\script\pptemplate
pl2bat.bat ..\..\blib\script\pptemplate
gmake[2]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Gen'
gmake[2]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/SourceFilter'
gmake[2]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/SourceFilter'
gmake[2]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core'
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" Types.pm.PL Types.pm 
Types.pm.PL: using typedef long long PDL_Indx
Extracting Types.pm
making PDL_B...
making PDL_S...
making PDL_US...
making PDL_L...
making PDL_IND...
making PDL_LL...
making PDL_F...
making PDL_D...
making PDL_CF...
making PDL_CD...
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" pdl.h.PL pdl.h 
Using new 64bit index support
adding...
#define PDL_U PDL_US
#define PDL_N PDL_IND
#define PDL_Q PDL_LL
#define PDL_G PDL_CF
#define PDL_C PDL_CD
Extracting pdl.h
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" pdlcore.h.PL pdlcore.h 
Extracting pdlcore.h
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" pdlsimple.h.PL pdlsimple.h 
Using new 64bit index support
adding...
#define PDL_U PDL_US
#define PDL_N PDL_IND
#define PDL_Q PDL_LL
#define PDL_G PDL_CF
#define PDL_C PDL_CD
Extracting pdlsimple.h (with variable substitutions)
Running Mkbootstrap for Core ()
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 644 "Core.bs"
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- Core.bs ..\..\blib\arch\auto\PDL\Core\Core.bs 644
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "C:\berrybrew\5.28.0_64_PDL\perl\lib\ExtUtils/xsubpp"  -typemap C:\berrybrew\5.28.0_64_PDL\perl\lib\ExtUtils\typemap -typemap C:\BERRYB~1\528~1.0_6\data\.cpanm\work\1618819138.29432\PDL-2.038\Basic\Core\typemap  Core.xs > Core.xsc
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e mv -- Core.xsc Core.c
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "-I..\..\blib\arch" "-I..\..\blib\lib" pdldataswitch.c.PL pdldataswitch.c 
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "-I..\..\blib\arch" "-I..\..\blib\lib" pdlbadvalinit.c.PL pdlbadvalinit.c 
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 Core.c
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "-I..\..\blib\arch" "-I..\..\blib\lib" pdlcore.c.PL pdlcore.c 
Extracting pdlcore.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlcore.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlapi.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlhash.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlthread.c
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "-I..\..\blib\arch" "-I..\..\blib\lib" pdlconv.c.PL pdlconv.c 
Extracting pdlconv.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlconv.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlmagic.c
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe"  -e "require \"C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core/Dev.pm\"; PDL::Core::Dev->import(); exit genpp();" -- "pdlsections.g" > "pdlsections.c"
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core" -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"  -DPDL_PTHREAD  -DBADVAL_USENAN=0 -DBADVAL_PER_PDL=0 pdlsections.c
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Mksymlists \
     -e "Mksymlists('NAME'=>\"PDL::Core\", 'DLBASE' => 'Core', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
g++.exe Core.def -o ..\..\blib\arch\auto\PDL\Core\Core.xs.dll -mdll -s -L"C:\berrybrew\5.28.0_64_PDL\perl\lib\CORE" -L"C:\berrybrew\5.28.0_64_PDL\c\lib" Core.o pdlcore.o pdlapi.o pdlhash.o pdlthread.o pdlconv.o pdlmagic.o pdlsections.o   "C:\berrybrew\5.28.0_64_PDL\perl\lib\CORE\libperl528.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libpthread.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libm.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libmoldname.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libkernel32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libuser32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libgdi32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libwinspool.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libcomdlg32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libadvapi32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libshell32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libole32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\liboleaut32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libnetapi32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libuuid.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libws2_32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libmpr.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libwinmm.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libversion.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libodbc32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libodbccp32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libcomctl32.a" -Wl,--enable-auto-image-base
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 755 ..\..\blib\arch\auto\PDL\Core\Core.xs.dll
gmake[2]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core'
gmake[2]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Bad'
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "-I..\..\blib\arch" "-I..\..\blib\lib" "-MPDL::PP qw[PDL::Bad PDL::Bad Bad ]" bad.pd
Running Mkbootstrap for Bad ()
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 644 "Bad.bs"
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- Bad.bs ..\..\blib\arch\auto\PDL\Bad\Bad.bs 644
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "C:\berrybrew\5.28.0_64_PDL\perl\lib\ExtUtils/xsubpp"  -typemap C:\berrybrew\5.28.0_64_PDL\perl\lib\ExtUtils\typemap -typemap C:\BERRYB~1\528~1.0_6\data\.cpanm\work\1618819138.29432\PDL-2.038\Basic\Core\typemap.pdl  Bad.xs > Bad.xsc
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e mv -- Bad.xsc Bad.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core"   -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"   Bad.c
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Mksymlists \
     -e "Mksymlists('NAME'=>\"PDL::Bad\", 'DLBASE' => 'Bad', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
g++.exe Bad.def -o ..\..\blib\arch\auto\PDL\Bad\Bad.xs.dll -mdll -s -L"C:\berrybrew\5.28.0_64_PDL\perl\lib\CORE" -L"C:\berrybrew\5.28.0_64_PDL\c\lib" Bad.o   "C:\berrybrew\5.28.0_64_PDL\perl\lib\CORE\libperl528.a" -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32 -Wl,--enable-auto-image-base
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 755 ..\..\blib\arch\auto\PDL\Bad\Bad.xs.dll
gmake[2]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Bad'
gmake[2]: Entering directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Ops'
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "-I..\..\blib\arch" "-I..\..\blib\lib" "-MPDL::PP qw[PDL::Ops PDL::Ops Ops ]" ops.pd
Running Mkbootstrap for Ops ()
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 644 "Ops.bs"
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- Ops.bs ..\..\blib\arch\auto\PDL\Ops\Ops.bs 644
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" "C:\berrybrew\5.28.0_64_PDL\perl\lib\ExtUtils/xsubpp"  -typemap C:\berrybrew\5.28.0_64_PDL\perl\lib\ExtUtils\typemap -typemap C:\BERRYB~1\528~1.0_6\data\.cpanm\work\1618819138.29432\PDL-2.038\Basic\Core\typemap.pdl  Ops.xs > Ops.xsc
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Command -e mv -- Ops.xsc Ops.c
gcc -c  "-IC:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Core"   -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"2.038\" -DXS_VERSION=\"2.038\"  "-IC:\berrybrew\5.28.0_64_PDL\perl\lib\CORE"   Ops.c
ops.pd: In function 'pdl_log10_readdata':
ops.pd:311:17: warning: implicit declaration of function 'clog10f' [-Wimplicit-function-declaration]
 
                 ^      
ops.pd:311:17: warning: incompatible implicit declaration of built-in function 'clog10f'
ops.pd:304:17: warning: incompatible implicit declaration of built-in function 'clog10f'
       pp_line_numbers(__LINE__, 'if ( ' . $badcode . ' )
                 ^~~~~~~
ops.pd:312:17: warning: implicit declaration of function 'clog10' [-Wimplicit-function-declaration]
 $doc
                 ^     
ops.pd:312:17: warning: incompatible implicit declaration of built-in function 'clog10'
ops.pd:305:17: warning: incompatible implicit declaration of built-in function 'clog10'
       $SETBAD(b());
                 ^~~   
"C:\berrybrew\5.28.0_64_PDL\perl\bin\perl.exe" -MExtUtils::Mksymlists \
     -e "Mksymlists('NAME'=>\"PDL::Ops\", 'DLBASE' => 'Ops', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
g++.exe Ops.def -o ..\..\blib\arch\auto\PDL\Ops\Ops.xs.dll -mdll -s -L"C:\berrybrew\5.28.0_64_PDL\perl\lib\CORE" -L"C:\berrybrew\5.28.0_64_PDL\c\lib" Ops.o   "C:\berrybrew\5.28.0_64_PDL\perl\lib\CORE\libperl528.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libm.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libmoldname.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libkernel32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libuser32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libgdi32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libwinspool.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libcomdlg32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libadvapi32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libshell32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libole32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\liboleaut32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libnetapi32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libuuid.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libws2_32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libmpr.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libwinmm.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libversion.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libodbc32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libodbccp32.a" "C:\berrybrew\5.28.0_64_PDL\c\x86_64-w64-mingw32\lib\libcomctl32.a" -Wl,--enable-auto-image-base
Ops.o:Ops.c:(.text+0x6c1ab): undefined reference to `clog10'
Ops.o:Ops.c:(.text+0x6c49e): undefined reference to `clog10f'
Ops.o:Ops.c:(.text+0x6d1fe): undefined reference to `clog10f'
Ops.o:Ops.c:(.text+0x6d3ec): undefined reference to `clog10'
collect2.exe: error: ld returned 1 exit status
gmake[2]: *** [Makefile:463: ..\..\blib\arch\auto\PDL\Ops\Ops.xs.dll] Error 1
gmake[2]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic/Ops'
gmake[1]: *** [Makefile:524: subdirs] Error 2
gmake[1]: Leaving directory 'C:/BERRYB~1/528~1.0_6/data/.cpanm/work/1618819138.29432/PDL-2.038/Basic'
gmake: *** [Makefile:526: subdirs] Error 2
-> FAIL Installing PDL failed. See C:\BERRYB~1\528~1.0_6\data\.cpanm\work\1618819138.29432\build.log for details. Retry with --force to force install it.
```

**Comment by zmughal** [comment]
```
Ops.o:Ops.c:(.text+0x6c1ab): undefined reference to `clog10'
Ops.o:Ops.c:(.text+0x6c49e): undefined reference to `clog10f'
Ops.o:Ops.c:(.text+0x6d1fe): undefined reference to `clog10f'
Ops.o:Ops.c:(.text+0x6d3ec): undefined reference to `clog10'
```

These seem to be the interesting lines out of the build log (for the build so far).

Per <https://strawberryperl.com/release-notes/5.28.0.1-64bit.html>, the distribution includes 

> 22.   mingw-w64-crt-v5.0.2 | http://mingw-w64.sourceforge.net/


which does not define `clog10`/`clog10f` <https://github.com/mirror/mingw-w64/blob/v5.0.2/mingw-w64-headers/crt/complex.h>.

These functions do not appear until <https://github.com/mirror/mingw-w64/blob/v6.0.0/mingw-w64-headers/crt/complex.h> which is included in <https://strawberryperl.com/release-notes/5.30.3.1-64bit.html>:

> 29.  mingw-w64-crt-v6.0.0 | http://mingw-w64.sourceforge.net/

**Comment by zmughal** [comment]
The existence of `clog10`/`clog10f` are done through Devel::CheckLib <https://github.com/PDLPorters/pdl/blob/2.038/Basic/Core/Dev.pm#L837>. @mohawk2, any ideas why they are still being generated?

Of note, these are not C standard functions and are GNU extensions <https://man7.org/linux/man-pages/man3/clog10.3.html> (but are reserved for future use by the standard).

**Comment by mohawk2** [comment]
This sounds similar to a problem on older FreeBSD and CentOS, where Devel::CheckLib (as used in the `got_complex_version` function in PDL::Core::Dev) was giving false positives. Possibly it needs further updating to catch this case.

**Comment by mohawk2** [comment]
@kmx Can you shed any light on this Strawberry Perl problem? Specifically for MinGW-crt-5.0.2.

**Comment by zmughal** [comment]
This should be addressed by <https://github.com/PDLPorters/pdl/commit/d87c9ff81dad56cc743a4ceb5fb9d1d30270afc6> available in PDL ≥ v2.041.

**Comment by shawnlaffan** [comment]
PDL 2.041 compiles and tests for me on Strawberry Perl 5.28.0.

Thanks for the fix.
